### PR TITLE
Stanage: initial GPU, CUDA and cuDNN docs (for A100 nodes)

### PR DESCRIPTION
**Pull request by willfurnass**
Will need revising later when the H100 nodes come online.

### CODE REVIEW

File: stanagepilot/software/libs/cuda.rst [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+    ncu -o metrics ./myapplication <arguments>
+
+Nsight compute can capture many different metrics which are used to generate the different sections of the profiling report. The ``--set`` argument can be used to control which set of metrics and sections are captured. See the `Nsight Compute CLI Command Line Options <https://docs.nvidia.com/nsight-compute/NsightComputeCli/index.html#command-line-options-profile>`_ for further information. ::
+
+    ncu -o metrics --set full ./myapplication <arguments>
+
+Once this file has been downloaded to your local machine, it can be opened in ``ncu-ui``/``nv-nsight-cu`` via File > Open File > metrics.ncu-rep
+
+Profiling using nvprof
```
Comment (ptheywood) [inline stanagepilot/software/libs/cuda.rst:200]: `nvprof` is not supported on SM80+ devices (i.e. all present in Stanage), so either a simple sentence stating it is not supported and to use `nsys`/`ncu`, or just remove the section. 

> Visual Profiler and nvprof don’t support devices with compute capability 8.0 and higher. Next-gen tools NVIDIA Nsight Compute and NVIDIA Nsight Systems should be used instead.

[source: `CUDA Profilers User Guide: Profiler changes in CUDA 11.0`](https://docs.nvidia.com/cuda/profiler-users-guide/#changelog)

File: stanagepilot/gpu-computing.rst [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+which is significantly less than the amount of GPU RAM available on a single GPU.
+This can lead to issues where your session has insufficient CPU RAM to transfer data to and from the GPU.
+As such, it is recommended that you request enough CPU memory to communicate properly with the GPU:
+
+.. code-block:: sh
+
+   # NB Each NVIDIA A100 GPU has 40GB of RAM
+   srun --partition=gpu --gres=gpu:1 --mem=42G --pty bash
+
+The above will give you 2GB more CPU RAM than the 40GB of GPU RAM available on the NVIDIA A100.
```
Comment (ptheywood) [inline stanagepilot/gpu-computing.rst:37]:
```suggestion
.. code-block:: sh

   # NB Each NVIDIA A100 GPU has 80GB of RAM
   srun --partition=gpu --gres=gpu:1 --mem=82G --pty bash

The above will give you 2GB more CPU RAM than the 80GB of GPU RAM available on the NVIDIA A100.
```

An old nvidia-smi log has the GPUs as being the 80GiB variant `NVIDIA A100-SXM4-80GB`, `0MiB / 81920MiB`

File: stanagepilot/software/libs/cuda.rst [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+i.e. ``-gencode=arch=compute_80,code=compute_80``.
+
+The minimum specified virtual architecture must be less than or equal to the `Compute Capability <https://developer.nvidia.com/cuda-gpus>`_ of the GPU used to execute the code.
+
+At present, all GPUs in Stanage are NVIDIA Tesla A100 GPUs, which are Compute Capability 80.
+To build a CUDA application which targets just A100 GPUs, use the following ``-gencode`` arguments:
+
+.. code-block:: sh
+
+   nvcc filename.cu \
```
Comment (ptheywood) [inline stanagepilot/software/libs/cuda.rst:140]: This is correct, but will only work with CUDA >= 11.0 which introdcued SM 80. 

[CUDA 11.0 Release Notes] (https://docs.nvidia.com/cuda/archive/11.0_GA/cuda-toolkit-release-notes/index.html#cuda-general-new-features).

So may need to add a condition around that (if CUDA 10.x is avialable as suggested in `Compiling the sample programs`), where embedding PTX for 70 via `-gencode=arch=compute_70,code=compute_70` would be the closest fit (but ideally everyone should use CUDA 11.0+ on these nodes).

File: stanagepilot/software/libs/cudnn.rst [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+cuDNN is part of the NVIDIA Deep Learning SDK.
+
+Usage
+-----
+
+.. warning::
+
+   **Only GPU-enabled nodes are able to run the cuDNN library.**
+
+.. See** :ref:`GPUComputing_bessemer` **for more information on how to request a GPU-enabled node for an interactive session or job submission.**
```
Comment (ptheywood) [inline stanagepilot/software/libs/cudnn.rst:26]:
```suggestion
.. See** :ref:`gpu_computing_stanage` **for more information on how to request a GPU-enabled node for an interactive session or job submission.**
```

I think that's the right xref.

File: stanagepilot/gpu-computing.rst (hunk 2 above)
Comment (willfurnass) [reply stanagepilot/gpu-computing.rst:37]: Doh!  I'd even mentioned they were the 80GB versions in an addition to the stanage-specs.rst page

File: stanagepilot/software/libs/cudnn.rst [hunk 5]
```diff
@@ -0,0 +1,58 @@
+.. _cudnn_stanage:
+
+cuDNN
+=====
+
+.. sidebar:: cuDNN
+
+   :Dependencies: CUDA, gcc
+   :URL: https://developer.nvidia.com/cudnn
+   :Documentation: https://developer.nvidia.com/cudnn
```
[the remark this answers is not in the document]
Comment (jkwmoore) [reply stanagepilot/software/libs/cudnn.rst:10]: Far as I can tell, these links are a muchness of a muchness and reference more or less the same content and have the same linkages.

>Or are we avoiding breaking links if structure changes in future?

I don't think I have enough info to know if that is the case but I wouldn't expect either link to get deleted by nvidia any time soon.

[the remark this answers is not in the document]
Comment (willfurnass) [reply stanagepilot/software/libs/cudnn.rst]: Seems https://developer.nvidia.com/cudnn now redirects to https://developer.nvidia.com so  I've switched to using the URL Carl found.

File: stanagepilot/software/libs/cudnn.rst [hunk 6]
```diff
@@ -0,0 +1,58 @@
+.. _cudnn_stanage:
+
+cuDNN
+=====
+
+.. sidebar:: cuDNN
+
+   :Dependencies: CUDA, gcc
+   :URL: https://developer.nvidia.com/cudnn
```
Comment (willfurnass) [inline stanagepilot/software/libs/cudnn.rst:9]:
```suggestion
   :URL: https://docs.nvidia.com/cudnn
```

File: stanagepilot/software/libs/cudnn.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (willfurnass) [reply stanagepilot/software/libs/cudnn.rst:10]: Huh, you're right.  Dunno what happened earlier.  Reverted.

File: stanagepilot/software/development/index.rst [hunk 7]
```diff
@@ -16,4 +18,4 @@ Developing on Stanage
         :maxdepth: 1
```
Comment (jkwmoore) [inline stanagepilot/software/development/index.rst:18]: Might be time to enable this bit as the toctree?

Comment (jkwmoore) [reply stanagepilot/software/development/index.rst:18]: i.e. LN14 - 21

**Review by jkwmoore — CHANGES_REQUESTED** [review]
Added some suggestions - hopefully they're visible now.

**Pull request merged**